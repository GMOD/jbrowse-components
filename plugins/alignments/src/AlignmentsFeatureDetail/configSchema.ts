import { ConfigurationSchema } from '@jbrowse/core/configuration'

export const configSchema = ConfigurationSchema('AlignmentsFeatureWidget', {})
