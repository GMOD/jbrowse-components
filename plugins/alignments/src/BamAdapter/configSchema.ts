import { ConfigurationSchema } from '@gmod/jbrowse-core/configuration'
import { types } from 'mobx-state-tree'

export default types.late(() =>
  ConfigurationSchema(
    'BamAdapter',
    {
      bamLocation: {
        type: 'fileLocation',
        defaultValue: { uri: '/path/to/my.bam' },
      },
      index: ConfigurationSchema('BamIndex', {
        indexType: {
          model: types.enumeration('IndexType', ['BAI', 'CSI']),
          type: 'stringEnum',
          defaultValue: 'BAI',
        },
        location: {
          type: 'fileLocation',
          defaultValue: { uri: '/path/to/my.bam.bai' },
        },
      }),
      chunkSizeLimit: {
        type: 'number',
        defaultValue: 100000000,
      },
      fetchSizeLimit: {
        type: 'number',
        defaultValue: 500000000,
      },
      sequenceAdapter: {
        type: 'frozen',
        defaultValue: null,
      },
    },
    { explicitlyTyped: true },
  ),
)
