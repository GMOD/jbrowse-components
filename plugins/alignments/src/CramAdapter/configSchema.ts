import PluginManager from '@jbrowse/core/PluginManager'
import { ConfigurationSchema } from '@jbrowse/core/configuration'

export default (pluginManager: PluginManager) => {
  const { types } = pluginManager.lib['mobx-state-tree']
  return types.late(() =>
    ConfigurationSchema(
      'CramAdapter',
      {
        cramLocation: {
          type: 'fileLocation',
          defaultValue: { uri: '/path/to/my.cram' },
        },
        craiLocation: {
          type: 'fileLocation',
          defaultValue: { uri: '/path/to/my.cram.crai' },
        },
        sequenceAdapter: pluginManager.pluggableConfigSchemaType('adapter'),
      },
      { explicitlyTyped: true },
    ),
  )
}
