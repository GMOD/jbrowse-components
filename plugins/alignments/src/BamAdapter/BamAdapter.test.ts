import { toArray } from 'rxjs/operators'
import Adapter from './BamAdapter'
import configSchema from './configSchema'

test('adapter can fetch features from volvox.bam', async () => {
  const adapter = new Adapter(
    configSchema.create({
      bamLocation: {
        localPath: require.resolve('../../test_data/volvox-sorted.bam'),
      },
      index: {
        location: {
          localPath: require.resolve('../../test_data/volvox-sorted.bam.bai'),
        },
      },
    }),
  )

  const features = adapter.getFeatures({
    assemblyName: 'volvox',
    refName: 'ctgA',
    start: 0,
    end: 20000,
  })

  const featuresArray = await features.pipe(toArray()).toPromise()
  expect(featuresArray[0].get('refName')).toBe('ctgA')
  const featuresJsonArray = featuresArray.map(f => f.toJSON())
  expect(featuresJsonArray.length).toEqual(3809)
  expect(featuresJsonArray.slice(1000, 1010)).toMatchSnapshot()

  expect(adapter.refIdToName(0)).toBe('ctgA')
  expect(adapter.refIdToName(1)).toBe(undefined)

  expect(await adapter.hasDataForRefName('ctgA')).toBe(true)

  const adapterCSI = new Adapter(
    configSchema.create({
      bamLocation: {
        localPath: require.resolve('../../test_data/volvox-sorted.bam'),
      },
      index: {
        indexType: 'CSI',
        location: {
          localPath: require.resolve('../../test_data/volvox-sorted.bam.csi'),
        },
      },
    }),
  )

  const featuresCSI = adapterCSI.getFeatures({
    assemblyName: 'volvox',
    refName: 'ctgA',
    start: 0,
    end: 20000,
  })
  const featuresArrayCSI = await featuresCSI.pipe(toArray()).toPromise()
  const featuresJsonArrayCSI = featuresArrayCSI.map(f => f.toJSON())
  expect(featuresJsonArrayCSI).toEqual(featuresJsonArray)
})

test('test usage of BamSlightlyLazyFeature toJSON (used in the widget)', async () => {
  const adapter = new Adapter(
    configSchema.create({
      bamLocation: {
        localPath: require.resolve('../../test_data/volvox-sorted.bam'),
      },
      index: {
        location: {
          localPath: require.resolve('../../test_data/volvox-sorted.bam.bai'),
        },
        indexType: 'BAI',
      },
    }),
  )

  const features = adapter.getFeatures({
    assemblyName: 'volvox',
    refName: 'ctgA',
    start: 0,
    end: 100,
  })
  const featuresArray = await features.pipe(toArray()).toPromise()
  const f = featuresArray[0].toJSON()
  expect(f.refName).toBe('ctgA')
  expect(f.start).toBe(2)
  expect(f.end).toBe(102)
  expect(f.mismatches).not.toBeTruthy()
})

test('test usage of BamSlightlyLazyFeature for extended CIGAR', async () => {
  const adapter = new Adapter(
    configSchema.create({
      bamLocation: {
        localPath: require.resolve('../../test_data/extended_cigar.bam'),
      },
      index: {
        location: {
          localPath: require.resolve('../../test_data/extended_cigar.bam.bai'),
        },
        indexType: 'BAI',
      },
    }),
  )

  const features = adapter.getFeatures({
    assemblyName: 'hg19',
    refName: '1',
    start: 13260,
    end: 13340,
  })
  const featuresArray = await features.pipe(toArray()).toPromise()
  const f = featuresArray[0]
  expect(f.get('mismatches')).toMatchSnapshot()
})
