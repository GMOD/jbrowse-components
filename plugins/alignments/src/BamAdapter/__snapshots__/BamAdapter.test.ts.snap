// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adapter can fetch features from volvox.bam 1`] = `
Array [
  Object {
    "CIGAR": "100M",
    "MD": "22A32C15T12A15",
    "MQ": 37,
    "NM": 4,
    "X0": 1,
    "X1": 0,
    "XG": 0,
    "XM": 4,
    "XO": 0,
    "XT": "U",
    "clipPos": 0,
    "end": 5540,
    "flags": 16,
    "length_on_ref": 100,
    "name": "ctgA_5060_5540_1:0:0_3:1:0_15a3",
    "qual": "17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17",
    "refName": "ctgA",
    "score": 37,
    "seq": "TACACTGGTTCGGAGACGGTTCGTGACGAGCGCGCTATATGTCGGCATCTGCGCCGCATGAGCGGCCGCTGACCGGCGGCACGACTAATATAGTGCAAGA",
    "seq_length": 100,
    "start": 5440,
    "strand": -1,
    "template_length": 0,
    "type": "match",
    "uniqueId": "test-10158054",
  },
  Object {
    "CIGAR": "100M",
    "MD": "66G33",
    "MQ": 37,
    "NM": 1,
    "X0": 1,
    "X1": 0,
    "XG": 0,
    "XM": 1,
    "XO": 0,
    "XT": "U",
    "clipPos": 0,
    "end": 5541,
    "flags": 16,
    "length_on_ref": 100,
    "name": "ctgA_4973_5541_1:0:0_1:0:0_1569",
    "qual": "17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17",
    "refName": "ctgA",
    "score": 37,
    "seq": "ACACTGGTTCGGAGACGGTTCATGACGAGCGCGCTATATGTCGGCATCTGCGCCCCATGAGCGGCCCCTGTCCGGCGGCACGAATAATATAGTGCAAGAA",
    "seq_length": 100,
    "start": 5441,
    "strand": -1,
    "template_length": 0,
    "type": "match",
    "uniqueId": "test-10158322",
  },
  Object {
    "CIGAR": "100M",
    "MD": "25G23C1C24A23",
    "MQ": 37,
    "NM": 4,
    "X0": 1,
    "X1": 0,
    "XG": 0,
    "XM": 4,
    "XO": 0,
    "XT": "U",
    "clipPos": 0,
    "end": 5544,
    "flags": 0,
    "length_on_ref": 100,
    "name": "ctgA_5445_5947_3:1:0_2:0:0_2b9",
    "qual": "17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17",
    "refName": "ctgA",
    "score": 37,
    "seq": "CTGGTTCGGAGACGGTTCATGACGACCGCGCTATATGTCGGCATCTGCGTCGCATGAGCGGCCGCTGTCCGGCGGCTCGAATAATATAGTGCAAGAAAAA",
    "seq_length": 100,
    "start": 5444,
    "strand": 1,
    "template_length": 0,
    "type": "match",
    "uniqueId": "test-10158581",
  },
  Object {
    "CIGAR": "100M",
    "MD": "42C57",
    "MQ": 37,
    "NM": 1,
    "X0": 1,
    "X1": 0,
    "XG": 0,
    "XM": 1,
    "XO": 0,
    "XT": "U",
    "clipPos": 0,
    "end": 5553,
    "flags": 0,
    "length_on_ref": 100,
    "name": "ctgA_5454_5906_0:1:0_2:0:0_d8a",
    "qual": "17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17",
    "refName": "ctgA",
    "score": 37,
    "seq": "AGACGGTTCATGACGAGCGCGCTATATGTCGGCATCTGCGCCGCATGAGCGGCCGCTGTCCGGCGGCACGAATAATATAGTGCAAGAAAAACCGAAGACT",
    "seq_length": 100,
    "start": 5453,
    "strand": 1,
    "template_length": 0,
    "type": "match",
    "uniqueId": "test-10158847",
  },
  Object {
    "CIGAR": "100M",
    "MD": "49G11G38",
    "MQ": 37,
    "NM": 2,
    "X0": 1,
    "X1": 0,
    "XG": 0,
    "XM": 2,
    "XO": 0,
    "XT": "U",
    "clipPos": 0,
    "end": 5554,
    "flags": 0,
    "length_on_ref": 100,
    "name": "ctgA_5455_5964_2:0:0_2:0:0_23e2",
    "qual": "17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17",
    "refName": "ctgA",
    "score": 37,
    "seq": "GACGGTTCATGACGAGCGCGCTATATGTCGGCATCTGCGCCCCATGAGCCGCCGCTGTCCGACGGCACGAATAATATAGTGCAAGAAAAACCGAAGACTA",
    "seq_length": 100,
    "start": 5454,
    "strand": 1,
    "template_length": 0,
    "type": "match",
    "uniqueId": "test-12607066",
  },
  Object {
    "CIGAR": "100M",
    "MD": "22T13C63",
    "MQ": 37,
    "NM": 2,
    "X0": 1,
    "X1": 0,
    "XG": 0,
    "XM": 2,
    "XO": 0,
    "XT": "U",
    "clipPos": 0,
    "end": 5559,
    "flags": 16,
    "length_on_ref": 100,
    "name": "ctgA_5153_5559_3:0:0_1:1:0_11ac",
    "qual": "17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17",
    "refName": "ctgA",
    "score": 37,
    "seq": "TTCATGACGAGCGCGCTATATGACGGCATCTGCGCCGCATGAGCGGCCGCTGTCCGGCGGCACGAATAATATAGTGCAAGAAAAACCGAAGACTACGGTT",
    "seq_length": 100,
    "start": 5459,
    "strand": -1,
    "template_length": 0,
    "type": "match",
    "uniqueId": "test-12607328",
  },
  Object {
    "CIGAR": "100M",
    "MD": "35C3G60",
    "MQ": 37,
    "NM": 2,
    "X0": 1,
    "X1": 0,
    "XG": 0,
    "XM": 2,
    "XO": 0,
    "XT": "U",
    "clipPos": 0,
    "end": 5560,
    "flags": 0,
    "length_on_ref": 100,
    "name": "ctgA_5461_5986_1:1:0_3:0:0_200e",
    "qual": "17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17",
    "refName": "ctgA",
    "score": 37,
    "seq": "TCATGACGAGCGCGCTATATGTCGGCATCTGCGCCGCATCAGCGGCCGCTGTCCGGCGGCACGAATAATATAGTGCAAGAAAAACCGAAGACTACGGTTA",
    "seq_length": 100,
    "start": 5460,
    "strand": 1,
    "template_length": 0,
    "type": "match",
    "uniqueId": "test-12607590",
  },
  Object {
    "CIGAR": "100M",
    "MD": "100",
    "MQ": 37,
    "NM": 0,
    "X0": 1,
    "X1": 0,
    "XG": 0,
    "XM": 0,
    "XO": 0,
    "XT": "U",
    "clipPos": 0,
    "end": 5568,
    "flags": 0,
    "length_on_ref": 100,
    "name": "ctgA_5469_5932_0:0:0_2:0:0_6f8",
    "qual": "17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17",
    "refName": "ctgA",
    "score": 37,
    "seq": "AGCGCGCTATATGTCGGCATCTGCGCCCCATGAGCGGCCGCTGTCCGGCGGCACGAATAATATAGTGCAAGAAAAACCGAAGACTACGGTTATATATGAT",
    "seq_length": 100,
    "start": 5468,
    "strand": 1,
    "template_length": 0,
    "type": "match",
    "uniqueId": "test-12607851",
  },
  Object {
    "CIGAR": "100M",
    "MD": "52G43A3",
    "MQ": 37,
    "NM": 2,
    "X0": 1,
    "X1": 0,
    "XG": 0,
    "XM": 2,
    "XO": 0,
    "XT": "U",
    "clipPos": 0,
    "end": 5594,
    "flags": 16,
    "length_on_ref": 100,
    "name": "ctgA_5079_5594_1:0:0_2:0:0_d64",
    "qual": "17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17",
    "refName": "ctgA",
    "score": 37,
    "seq": "CCCATGAGCGGCCGCTGTCCGGCGGCACGAATAATATAGTGCAAGAAAAACCTAAGACTACGGTTATATATGATGGAACGGCCCTCACAGCATTCTCACA",
    "seq_length": 100,
    "start": 5494,
    "strand": -1,
    "template_length": 0,
    "type": "match",
    "uniqueId": "test-12608107",
  },
  Object {
    "CIGAR": "100M",
    "MD": "92C7",
    "MQ": 37,
    "NM": 1,
    "X0": 1,
    "X1": 0,
    "XG": 0,
    "XM": 1,
    "XO": 0,
    "XT": "U",
    "clipPos": 0,
    "end": 5596,
    "flags": 0,
    "length_on_ref": 100,
    "name": "ctgA_5497_6082_1:0:0_2:0:0_b2",
    "qual": "17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17",
    "refName": "ctgA",
    "score": 37,
    "seq": "CATGAGCGGCCGCTGTCCGGCGGCACGAATAATATAGTGCAAGAAAAACCGAAGACTACGGTTATATATGATGGAACGGCCCTCACAGCATTGTAACAGG",
    "seq_length": 100,
    "start": 5496,
    "strand": 1,
    "template_length": 0,
    "type": "match",
    "uniqueId": "test-12608367",
  },
]
`;

exports[`test usage of BamSlightlyLazyFeature for extended CIGAR 1`] = `
Array [
  Object {
    "base": "S2666",
    "cliplen": 2666,
    "length": 1,
    "start": 0,
    "type": "softclip",
  },
  Object {
    "base": "*",
    "length": 1,
    "start": 33,
    "type": "deletion",
  },
  Object {
    "base": "G",
    "length": 1,
    "qual": 93,
    "start": 40,
    "type": "mismatch",
  },
  Object {
    "base": "*",
    "length": 1,
    "start": 75,
    "type": "deletion",
  },
  Object {
    "base": "*",
    "length": 1,
    "start": 90,
    "type": "deletion",
  },
  Object {
    "base": "1",
    "length": 0,
    "start": 108,
    "type": "insertion",
  },
  Object {
    "base": "*",
    "length": 1,
    "start": 132,
    "type": "deletion",
  },
  Object {
    "base": "*",
    "length": 1,
    "start": 146,
    "type": "deletion",
  },
  Object {
    "base": "3",
    "length": 0,
    "start": 177,
    "type": "insertion",
  },
  Object {
    "base": "1",
    "length": 0,
    "start": 209,
    "type": "insertion",
  },
  Object {
    "base": "C",
    "length": 1,
    "qual": 93,
    "start": 216,
    "type": "mismatch",
  },
  Object {
    "base": "C",
    "length": 1,
    "qual": 93,
    "start": 222,
    "type": "mismatch",
  },
  Object {
    "base": "C",
    "length": 1,
    "qual": 93,
    "start": 228,
    "type": "mismatch",
  },
  Object {
    "base": "*",
    "length": 1,
    "start": 230,
    "type": "deletion",
  },
  Object {
    "base": "C",
    "length": 1,
    "qual": 93,
    "start": 235,
    "type": "mismatch",
  },
  Object {
    "base": "C",
    "length": 1,
    "qual": 93,
    "start": 241,
    "type": "mismatch",
  },
  Object {
    "base": "*",
    "length": 1,
    "start": 247,
    "type": "deletion",
  },
  Object {
    "base": "C",
    "length": 1,
    "qual": 93,
    "start": 248,
    "type": "mismatch",
  },
  Object {
    "base": "*",
    "length": 1,
    "start": 254,
    "type": "deletion",
  },
  Object {
    "base": "C",
    "length": 1,
    "qual": 93,
    "start": 255,
    "type": "mismatch",
  },
  Object {
    "base": "*",
    "length": 1,
    "start": 263,
    "type": "deletion",
  },
  Object {
    "base": "1",
    "length": 0,
    "start": 284,
    "type": "insertion",
  },
  Object {
    "base": "T",
    "length": 1,
    "qual": 93,
    "start": 289,
    "type": "mismatch",
  },
  Object {
    "base": "T",
    "length": 1,
    "qual": 93,
    "start": 290,
    "type": "mismatch",
  },
  Object {
    "base": "*",
    "length": 1,
    "start": 293,
    "type": "deletion",
  },
  Object {
    "base": "A",
    "length": 1,
    "qual": 93,
    "start": 294,
    "type": "mismatch",
  },
  Object {
    "base": "T",
    "length": 1,
    "qual": 93,
    "start": 296,
    "type": "mismatch",
  },
  Object {
    "base": "T",
    "length": 1,
    "qual": 93,
    "start": 302,
    "type": "mismatch",
  },
  Object {
    "base": "G",
    "length": 1,
    "qual": 93,
    "start": 308,
    "type": "mismatch",
  },
  Object {
    "base": "G",
    "length": 1,
    "qual": 93,
    "start": 314,
    "type": "mismatch",
  },
  Object {
    "base": "G",
    "length": 1,
    "qual": 93,
    "start": 320,
    "type": "mismatch",
  },
  Object {
    "base": "G",
    "length": 1,
    "qual": 93,
    "start": 326,
    "type": "mismatch",
  },
  Object {
    "base": "*",
    "length": 1,
    "start": 329,
    "type": "deletion",
  },
  Object {
    "base": "G",
    "length": 1,
    "qual": 93,
    "start": 333,
    "type": "mismatch",
  },
  Object {
    "base": "1",
    "length": 0,
    "start": 339,
    "type": "insertion",
  },
  Object {
    "base": "1",
    "length": 0,
    "start": 345,
    "type": "insertion",
  },
  Object {
    "base": "2",
    "length": 0,
    "start": 352,
    "type": "insertion",
  },
  Object {
    "base": "*",
    "length": 1,
    "start": 375,
    "type": "deletion",
  },
  Object {
    "base": "G",
    "length": 1,
    "qual": 93,
    "start": 384,
    "type": "mismatch",
  },
  Object {
    "base": "67",
    "length": 0,
    "start": 403,
    "type": "insertion",
  },
  Object {
    "base": "1",
    "length": 0,
    "start": 427,
    "type": "insertion",
  },
  Object {
    "base": "1",
    "length": 0,
    "start": 433,
    "type": "insertion",
  },
  Object {
    "base": "*",
    "length": 1,
    "start": 439,
    "type": "deletion",
  },
  Object {
    "base": "T",
    "length": 1,
    "qual": 93,
    "start": 491,
    "type": "mismatch",
  },
  Object {
    "base": "G",
    "length": 1,
    "qual": 93,
    "start": 621,
    "type": "mismatch",
  },
  Object {
    "base": "C",
    "length": 1,
    "qual": 93,
    "start": 622,
    "type": "mismatch",
  },
  Object {
    "base": "7",
    "length": 0,
    "start": 626,
    "type": "insertion",
  },
  Object {
    "base": "1",
    "length": 0,
    "start": 627,
    "type": "insertion",
  },
  Object {
    "base": "G",
    "length": 1,
    "qual": 93,
    "start": 800,
    "type": "mismatch",
  },
  Object {
    "base": "G",
    "length": 1,
    "qual": 93,
    "start": 802,
    "type": "mismatch",
  },
  Object {
    "base": "G",
    "length": 1,
    "qual": 93,
    "start": 803,
    "type": "mismatch",
  },
  Object {
    "base": "C",
    "length": 1,
    "qual": 93,
    "start": 804,
    "type": "mismatch",
  },
  Object {
    "base": "*",
    "length": 1,
    "start": 807,
    "type": "deletion",
  },
  Object {
    "base": "G",
    "length": 1,
    "qual": 93,
    "start": 808,
    "type": "mismatch",
  },
  Object {
    "base": "C",
    "length": 1,
    "qual": 93,
    "start": 809,
    "type": "mismatch",
  },
  Object {
    "base": "C",
    "length": 1,
    "qual": 93,
    "start": 814,
    "type": "mismatch",
  },
  Object {
    "base": "4",
    "length": 0,
    "start": 818,
    "type": "insertion",
  },
  Object {
    "base": "C",
    "length": 1,
    "qual": 93,
    "start": 820,
    "type": "mismatch",
  },
  Object {
    "base": "4",
    "length": 0,
    "start": 822,
    "type": "insertion",
  },
  Object {
    "base": "8",
    "length": 0,
    "start": 828,
    "type": "insertion",
  },
  Object {
    "base": "C",
    "length": 1,
    "qual": 93,
    "start": 828,
    "type": "mismatch",
  },
  Object {
    "base": "G",
    "length": 1,
    "qual": 93,
    "start": 883,
    "type": "mismatch",
  },
  Object {
    "base": "47",
    "length": 0,
    "start": 889,
    "type": "insertion",
  },
  Object {
    "base": "A",
    "length": 1,
    "qual": 93,
    "start": 896,
    "type": "mismatch",
  },
  Object {
    "base": "G",
    "length": 1,
    "qual": 93,
    "start": 926,
    "type": "mismatch",
  },
  Object {
    "base": "T",
    "length": 1,
    "qual": 93,
    "start": 930,
    "type": "mismatch",
  },
  Object {
    "base": "T",
    "length": 1,
    "qual": 93,
    "start": 933,
    "type": "mismatch",
  },
  Object {
    "base": "C",
    "length": 1,
    "qual": 93,
    "start": 935,
    "type": "mismatch",
  },
  Object {
    "base": "1",
    "length": 0,
    "start": 968,
    "type": "insertion",
  },
  Object {
    "base": "C",
    "length": 1,
    "qual": 93,
    "start": 1001,
    "type": "mismatch",
  },
  Object {
    "base": "T",
    "length": 1,
    "qual": 93,
    "start": 1016,
    "type": "mismatch",
  },
  Object {
    "base": "G",
    "length": 1,
    "qual": 39,
    "start": 1018,
    "type": "mismatch",
  },
  Object {
    "base": "A",
    "length": 1,
    "qual": 93,
    "start": 1034,
    "type": "mismatch",
  },
  Object {
    "base": "1",
    "length": 0,
    "start": 1063,
    "type": "insertion",
  },
  Object {
    "base": "2",
    "length": 0,
    "start": 1068,
    "type": "insertion",
  },
  Object {
    "base": "G",
    "length": 1,
    "qual": 93,
    "start": 1074,
    "type": "mismatch",
  },
  Object {
    "base": "G",
    "length": 1,
    "qual": 93,
    "start": 1078,
    "type": "mismatch",
  },
  Object {
    "base": "C",
    "length": 1,
    "qual": 93,
    "start": 1112,
    "type": "mismatch",
  },
  Object {
    "base": "G",
    "length": 1,
    "qual": 93,
    "start": 1133,
    "type": "mismatch",
  },
  Object {
    "base": "C",
    "length": 1,
    "qual": 93,
    "start": 1153,
    "type": "mismatch",
  },
  Object {
    "base": "A",
    "length": 1,
    "qual": 93,
    "start": 1166,
    "type": "mismatch",
  },
  Object {
    "base": "A",
    "length": 1,
    "qual": 93,
    "start": 1390,
    "type": "mismatch",
  },
  Object {
    "base": "G",
    "length": 1,
    "qual": 93,
    "start": 1408,
    "type": "mismatch",
  },
  Object {
    "base": "G",
    "length": 1,
    "qual": 93,
    "start": 1456,
    "type": "mismatch",
  },
  Object {
    "base": "G",
    "length": 1,
    "qual": 93,
    "start": 1507,
    "type": "mismatch",
  },
  Object {
    "base": "T",
    "length": 1,
    "qual": 93,
    "start": 1541,
    "type": "mismatch",
  },
  Object {
    "base": "T",
    "length": 1,
    "qual": 93,
    "start": 1564,
    "type": "mismatch",
  },
  Object {
    "base": "A",
    "length": 1,
    "qual": 93,
    "start": 1771,
    "type": "mismatch",
  },
  Object {
    "base": "C",
    "length": 1,
    "qual": 93,
    "start": 2197,
    "type": "mismatch",
  },
  Object {
    "base": "A",
    "length": 1,
    "qual": 93,
    "start": 2382,
    "type": "mismatch",
  },
  Object {
    "base": "A",
    "length": 1,
    "qual": 93,
    "start": 2782,
    "type": "mismatch",
  },
  Object {
    "base": "G",
    "length": 1,
    "qual": 93,
    "start": 3115,
    "type": "mismatch",
  },
  Object {
    "base": "G",
    "length": 1,
    "qual": 93,
    "start": 3117,
    "type": "mismatch",
  },
  Object {
    "base": "T",
    "length": 1,
    "qual": 93,
    "start": 3301,
    "type": "mismatch",
  },
  Object {
    "base": "C",
    "length": 1,
    "qual": 93,
    "start": 3326,
    "type": "mismatch",
  },
  Object {
    "base": "*",
    "length": 2,
    "start": 3656,
    "type": "deletion",
  },
  Object {
    "base": "A",
    "length": 1,
    "qual": 93,
    "start": 3895,
    "type": "mismatch",
  },
  Object {
    "base": "*",
    "length": 1,
    "start": 3957,
    "type": "deletion",
  },
  Object {
    "base": "C",
    "length": 1,
    "qual": 93,
    "start": 3979,
    "type": "mismatch",
  },
  Object {
    "base": "5",
    "length": 0,
    "start": 4455,
    "type": "insertion",
  },
  Object {
    "base": "C",
    "length": 1,
    "qual": 93,
    "start": 4672,
    "type": "mismatch",
  },
  Object {
    "base": "G",
    "length": 1,
    "qual": 93,
    "start": 4698,
    "type": "mismatch",
  },
  Object {
    "base": "G",
    "length": 1,
    "qual": 93,
    "start": 4906,
    "type": "mismatch",
  },
  Object {
    "base": "G",
    "length": 1,
    "qual": 93,
    "start": 4929,
    "type": "mismatch",
  },
  Object {
    "base": "G",
    "length": 1,
    "qual": 93,
    "start": 5117,
    "type": "mismatch",
  },
  Object {
    "base": "G",
    "length": 1,
    "qual": 93,
    "start": 5210,
    "type": "mismatch",
  },
  Object {
    "base": "T",
    "length": 1,
    "qual": 93,
    "start": 5273,
    "type": "mismatch",
  },
  Object {
    "base": "T",
    "length": 1,
    "qual": 93,
    "start": 5816,
    "type": "mismatch",
  },
  Object {
    "base": "T",
    "length": 1,
    "qual": 93,
    "start": 5819,
    "type": "mismatch",
  },
  Object {
    "base": "1",
    "length": 0,
    "start": 5903,
    "type": "insertion",
  },
  Object {
    "base": "G",
    "length": 1,
    "qual": 93,
    "start": 5905,
    "type": "mismatch",
  },
  Object {
    "base": "G",
    "length": 1,
    "qual": 93,
    "start": 5921,
    "type": "mismatch",
  },
  Object {
    "base": "A",
    "length": 1,
    "qual": 93,
    "start": 5955,
    "type": "mismatch",
  },
  Object {
    "base": "C",
    "length": 1,
    "qual": 93,
    "start": 6067,
    "type": "mismatch",
  },
  Object {
    "base": "G",
    "length": 1,
    "qual": 93,
    "start": 6102,
    "type": "mismatch",
  },
  Object {
    "base": "T",
    "length": 1,
    "qual": 93,
    "start": 6297,
    "type": "mismatch",
  },
  Object {
    "base": "C",
    "length": 1,
    "qual": 93,
    "start": 6377,
    "type": "mismatch",
  },
  Object {
    "base": "T",
    "length": 1,
    "qual": 93,
    "start": 6533,
    "type": "mismatch",
  },
  Object {
    "base": "A",
    "length": 1,
    "qual": 93,
    "start": 6570,
    "type": "mismatch",
  },
  Object {
    "base": "C",
    "length": 1,
    "qual": 93,
    "start": 6806,
    "type": "mismatch",
  },
  Object {
    "base": "G",
    "length": 1,
    "qual": 93,
    "start": 6808,
    "type": "mismatch",
  },
  Object {
    "base": "G",
    "length": 1,
    "qual": 93,
    "start": 6855,
    "type": "mismatch",
  },
  Object {
    "base": "S1",
    "cliplen": 1,
    "length": 1,
    "start": 7531,
    "type": "softclip",
  },
]
`;
